design/acme-orders-challenges-crd.md

```diff
@@ -0,0 +1,388 @@
+
+	// Processing signifies that an ACME resource is being processed by the server.
+	// If an Order is marked 'Processing', the validations for that Order are currently being processed.
+	// This is a transient state.
+	Processing State = "processing"
+
+	// Failed signifies that an ACME resource has failed for some reason.
+	// If an Order is marked 'Failed', one of its validations may have failed for some reason.
+	// This is a final state.
+	Failed State = "failed"
+
+	// Expired signifies that an ACME resource has expired.
+	// If an Order is marked 'Expired', one of its validations may have expired or the Order itself.
+	// This is a final state.
+	Expired State = "expired"
+)
+```
+
+##### Challenge
+
+```go
+// +genclient
+// +k8s:openapi-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +kubebuilder:resource:path=challenges
+type Challenge struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   ChallengeSpec   `json:"spec"`
+	Status ChallengeStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ChallengeList is a list of Challenges
+type ChallengeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []Challenge `json:"items"`
+}
+
+type ChallengeSpec struct {
+	AuthzURL string `json:"authzURL"`
+	Type     string `json:"type"`
+	URL      string `json:"url"`
+	DNSName  string `json:"dnsName"`
+	Token    string `json:"token"`
+	Key      string `json:"key"`
+	Wildcard bool   `json:"wildcard"`
+
+	// Config specifies the solver configuration for this challenge.
+	Config SolverConfig `json:"config"`
+
+	// IssuerRef references a properly configured ACME-type Issuer which should
+	// be used to create this Challenge.
+	// If the Issuer does not exist, processing will be retried.
+	// If the Issuer is not an 'ACME' Issuer, an error will be returned and the
+	// Challenge will be marked as failed.
+	IssuerRef ObjectReference `json:"issuerRef"`
+}
+
+type ChallengeStatus struct {
+	Presented bool   `json:"presented"`
+	Reason    string `json:"reason"`
+	State     State  `json:"state"`
+}
+```
+
+#### Changes to existing types
+
+The scope of changes to existing types as been purposely kept as small as possible.
+
+We do however require a change to the `status` field of the Certificate resource
+type. This should not be too problematic for users, as the cert-manager controller
+should be able to tolerate this change seamlessly.
+
+```go
+// CertificateStatus defines the observed state of Certificate
+type CertificateStatus struct {
+	Conditions      []CertificateCondition `json:"conditions,omitempty"`
+    ACME            *CertificateACMEStatus `json:"acme,omitempty"`
+    /////// NEW FIELD ///////
+	LastFailureTime *metav1.Time           `json:"lastFailureTime,omitempty"`
+}
+
+// CertificateACMEStatus holds the status for an ACME issuer
+type CertificateACMEStatus struct {
```

Comment:
This structure has now been removed altogether from the implementation in #788 as it could cause race conditions where our controller has not observed changes to the Certificate resource, resulting in duplicate Orders being created.

We now rely on labels/selectors to 'discover' orders for a given challenge. We also use the hash of the order's spec as part of its own name, in order to have a deterministic name that we can attempt to create and have rejected if the resource already exists due to a cache race condition.

There is some discussion on a similar issue upstream here: https://github.com/kubernetes/kubernetes/issues/22061

Quoting: `I'd like to do the same for creation by making pod names deterministic.` - doing this has worked well in this instance!